runtime-modules/bounty/src/lib.rs

```diff
@@ -1179,6 +1206,17 @@ decl_module! {
                 }
             }
 
+
+            //Transferers oracle cherry to the oracle
+            bounty_oracle_manager
```

Comment:
Potentially unsafe operator.

Reply:
I did a balance check, so the total cherries are taken into account.
```rust
#[weight = WeightInfoBounty::<T>::cancel_bounty_by_member().max(WeightInfoBounty::<T>::cancel_bounty_by_council())]
        pub fn cancel_bounty(origin, creator: BountyActor<MemberId<T>>, bounty_id: T::BountyId) {

            let bounty_creator_manager = BountyActorManager::<T>::ensure_bounty_actor_manager(
                origin,
                creator.clone(),
            )?;

            let bounty = Self::ensure_bounty_exists(&bounty_id)?;

            bounty_creator_manager.validate_actor(&bounty.creation_params.creator)?;

            let current_bounty_stage = Self::get_bounty_stage(&bounty);

            Self::ensure_bounty_stage_for_canceling(current_bounty_stage)?;

            let bounty_account_id = Self::bounty_account_id(bounty_id);
            let balance = balances::Module::<T>::usable_balance(&bounty_account_id);

            ensure!(balance >= bounty.creation_params.cherry + bounty.creation_params.oracle_cherry, 
                Error::<T>::NoBountyBalanceToCherryWithdrawal);

            //
            // == MUTATION SAFE ==
            //

            Self::return_bounty_cherry_to_creator(bounty_id, &bounty)?;
            Self::return_bounty_oracle_cherry_to_creator(bounty_id, &bounty)?;

            Self::remove_bounty(&bounty_id);

            Self::deposit_event(RawEvent::BountyCanceled(bounty_id, creator));

        } 
```

Reply:
I found also in the substrate docs, a transaction example with "mutate", but that's what is being done inside the return_bounty_cherry functions
```rust
    let sender = ensure_signed(origin)?;

    Accounts::<T>::mutate(&sender, |bal| {
      *bal = bal.saturating_sub(amount);
      });
      Accounts::<T>::mutate(&to, |bal| {
        *bal = bal.saturating_add(amount);
        });

    Self::deposit_event(Event::<T>::Transferred(sender, to, amount))
    Ok(().into())
```

Reply:
The `mutation mark` issue was resolved correctly. I would add a comment describing the connection between the check and two unsafe mutations after the mark to prevent it from incorrectly modifying by subsequent coding.

Reply:
I'm not sure you got the correct code snippet. We don't pass the origin for the `transfer` method.

Reply:
> I'm not sure you got the correct code snippet. We don't pass the origin for the transfer method.

I know what I mean was the use of Accounts::<T>::mutate, which It's safe to use in opposition of Self::return_bounty_cherry_to_creator(bounty_id, &bounty)?;

Reply:
Are you suggesting using `Accounts::<T>::mutate` directly?

Reply:
Well It has the advantage of being more clear, I can see in one little code snippet what is going on under the hood,
also doesn't terminate the execution in the middle like "?" does. 
If we have many mutations I would do this way, but with only two and to maintain consistency, I think we should keep the  Self::return_bounty_cherry_to_creator functions

Reply:
We can't rely on the `balances` pallet inner structure - because it could change at any moment by Parity's redesign. So, we use public API only.

Reply:
The oracle cherry is withdrawn in the following points:

- cancel_bounty
- veto_bounty 
- withdraw_funding
- withdraw_work_entrant_funds

What do you think of the withdraw_work_entrant_funds? 

```rust
       pub fn withdraw_work_entrant_funds(
            origin,
            member_id: MemberId<T>,
            bounty_id: T::BountyId,
            entry_id: T::EntryId,
        ) {
            let controller_account_id =
                T::Membership::ensure_member_controller_account_origin(origin, member_id)?;

            let bounty = Self::ensure_bounty_exists(&bounty_id)?;

            let current_bounty_stage = Self::get_bounty_stage(&bounty);

            // Ensure withdrawal for successful or failed bounty.
            ensure!(
                current_bounty_stage == BountyStage::FailedBountyWithdrawal ||
                current_bounty_stage == BountyStage::SuccessfulBountyWithdrawal,
                Self::unexpected_bounty_stage_error(current_bounty_stage)
            );

            let entry = Self::ensure_work_entry_exists(&entry_id)?;
        
            let bounty_account_id = Self::bounty_account_id(bounty_id);
            let balance = balances::Module::<T>::usable_balance(&bounty_account_id);

            BountyActorManager::<T>::get_bounty_actor_manager(
                bounty.creation_params.creator.clone(),
            )?;
            BountyActorManager::<T>::get_bounty_actor_manager(
                bounty.creation_params.oracle.clone(),
            )?;
      
            //
            // == MUTATION SAFE ==
            //

            // Claim the winner reward.
            if let Some(OracleWorkEntryJudgment::Winner { reward }) = entry.oracle_judgment_result {
                //Checks if there is suffiicient funds to safely perform cherry  and oracle cherry withdrawal
                ensure!(balance >= reward + bounty.creation_params.oracle_cherry, 
                    Error::<T>::NoBountyBalanceToCherryWithdrawal);

                Self::transfer_funds_from_bounty_account(
                    &controller_account_id,
                    bounty_id,
                    reward
                )?;
            }

            // Unstake the full work entry state.
            T::StakingHandler::unlock(&entry.staking_account_id);

            // Delete the work entry record from the storage.
            Self::remove_work_entry(&bounty_id, &entry_id);

            // Fire an event.
            Self::deposit_event(RawEvent::WorkEntrantFundsWithdrawn(bounty_id, entry_id, member_id));

            // Remove the bounty in case of the last withdrawal operation.
            if Self::withdrawal_completed(&current_bounty_stage, &bounty_id) {
                match bounty.milestone{
                    BountyMilestone::JudgmentSubmitted{successful_bounty:_} => {
                        //The oracle must receive the orracle cherry because there was a judgement.
                        Self::withdraw_oracle_cherry_to_oracle(bounty_id, &bounty)?
                    },
                    _ => {
                        //If stage is FailedBountyWithdrawal must be because of 
                        //founding period expired with insuficient funds
                        //working period expired with no active entries
                        //judgement period expired with no judge submissions
                        //In the above 3 cases the oracle cherry must return to the creator.
                        Self::return_bounty_oracle_cherry_to_creator(bounty_id, &bounty)?              
                    }                    
                };

                Self::remove_bounty(&bounty_id);
            }
        }
    }
```
At the end I check all conditions the oracle or the creator should receive the oracle cherry.
There are also checks to guarantee mutation safe, the  conditions at the end are already checked by this
```rust             
ensure!(
                current_bounty_stage == BountyStage::FailedBountyWithdrawal ||
                current_bounty_stage == BountyStage::SuccessfulBountyWithdrawal,
                Self::unexpected_bounty_stage_error(current_bounty_stage)
            );
```

Reply:
Here are withdraw_funding, it's similar, I can make more checks but it will be a little redundant, but if the stage logic in the other functions change this will fail 

```rust       
pub fn withdraw_funding(
            origin,
            funder: BountyActor<MemberId<T>>,
            bounty_id: T::BountyId,
        ) {
            let bounty_funder_manager = BountyActorManager::<T>::ensure_bounty_actor_manager(
                origin,
                funder.clone(),
            )?;

            let bounty = Self::ensure_bounty_exists(&bounty_id)?;

            let current_bounty_stage = Self::get_bounty_stage(&bounty);
            Self::ensure_bounty_stage(current_bounty_stage, BountyStage::FailedBountyWithdrawal)?;

            ensure!(
                <BountyContributions<T>>::contains_key(&bounty_id, &funder),
                Error::<T>::NoBountyContributionFound,
            );

            let funding_amount = <BountyContributions<T>>::get(&bounty_id, &funder);
            let cherry_fraction = Self::get_cherry_fraction_for_member(&bounty, funding_amount);

            let withdrawal_amount = funding_amount + cherry_fraction;
          
            let bounty_account_id = Self::bounty_account_id(bounty_id);
            let balance = balances::Module::<T>::usable_balance(&bounty_account_id);

            BountyActorManager::<T>::get_bounty_actor_manager(
                bounty.creation_params.creator.clone(),
            )?;
            BountyActorManager::<T>::get_bounty_actor_manager(
                bounty.creation_params.oracle.clone(),
            )?;

            //Checks if there is suffiicient funds to safely perform cherry  and oracle cherry withdrawal
            ensure!(balance >= withdrawal_amount + bounty.creation_params.oracle_cherry, 
                Error::<T>::NoBountyBalanceToCherryWithdrawal);

            //
            // == MUTATION SAFE ==
            //

            bounty_funder_manager.transfer_funds_from_bounty_account(bounty_id, withdrawal_amount)?;

            <BountyContributions<T>>::remove(&bounty_id, &funder);

            Self::deposit_event(RawEvent::BountyFundingWithdrawal(bounty_id, funder));
            
            if Self::withdrawal_completed(&current_bounty_stage, &bounty_id) {

                match bounty.milestone{
                    BountyMilestone::JudgmentSubmitted{successful_bounty:_} => {
                        //The oracle must receive the orracle cherry because there was a judgement.
                        Self::withdraw_oracle_cherry_to_oracle(bounty_id, &bounty)?
                    },
                    _ => {
                        //If stage is FailedBountyWithdrawal must be because of 
                        //founding period expired with insuficient funds
                        //working period expired with no active entries
                        //judgement period expired with no judge submissions
                        //In the above 3 cases the oracle cherry must return to the creator.
                        Self::return_bounty_oracle_cherry_to_creator(bounty_id, &bounty)?              
                    }                    
                };

                Self::remove_bounty(&bounty_id);
            }
        }
```

Reply:
submit_oracle_judgment doesn't withdraw oracle cherry anymore, this happens now at the end of the last withdrawal funds or in cancelling/veto events

Reply:
It seems unfair to the oracle to make it dependent on the funds' withdrawal from other parties that could never happen. The previous point of reward (`submit_oracle_judgment` extrinsic) looked better. The `JudgmentSubmitted` milestone is a correct condition to return the cherry. However, we have the bounty stage as the calculated object containing all the information about the bounty's current state. We should use only this object to query about allowed actions. BountyMilestone is a 'facts container for storage'. If the bounty stage info is not complete - please add required data (probably, `judgment_submitted: bool` in the `FailedBountyWithdrawal` stage).